test/e2e/tracing_test.go

```diff
@@ -110,10 +112,171 @@ var _ = Describe("Tracing", func() {
+				g.Expect(k8sClient.Get(ctx, key, &pipeline)).To(Succeed())
+				return pipeline.Status.HasCondition(telemetryv1alpha1.TracePipelineRunning)
+			}, timeout, interval).Should(BeTrue())
+
+			Expect(kitk8s.CreateObjects(ctx, k8sClient, pipeline2Objects...)).Should(Succeed())
+			Eventually(func(g Gomega) bool {
+				var pipeline telemetryv1alpha1.TracePipeline
+				key := types.NamespacedName{Name: "pipeline-2"}
+				g.Expect(k8sClient.Get(ctx, key, &pipeline)).To(Succeed())
+				return pipeline.Status.HasCondition(telemetryv1alpha1.TracePipelineRunning)
+			}, timeout, interval).Should(BeTrue())
+
+			Expect(kitk8s.CreateObjects(ctx, k8sClient, pipeline3Objects...)).Should(Succeed())
+			Eventually(func(g Gomega) bool {
+				var pipeline telemetryv1alpha1.TracePipeline
+				key := types.NamespacedName{Name: "pipeline-3"}
+				g.Expect(k8sClient.Get(ctx, key, &pipeline)).To(Succeed())
+				return pipeline.Status.HasCondition(telemetryv1alpha1.TracePipelineRunning)
+			}, timeout, interval).Should(BeTrue())
+
+			Expect(kitk8s.CreateObjects(ctx, k8sClient, pipeline4Objects...)).Should(Succeed())
+			time.Sleep(10 * time.Second) // wait for reconciliation
+			var pipeline telemetryv1alpha1.TracePipeline
+			key := types.NamespacedName{Name: "pipeline-4"}
+			Expect(k8sClient.Get(ctx, key, &pipeline)).To(Succeed())
+			// We allow only 3 piplines, pipeline-4 should not be running
+			Expect(!pipeline.Status.HasCondition(telemetryv1alpha1.TracePipelineRunning))
+
+			Expect(kitk8s.DeleteObjects(ctx, k8sClient, pipeline1Objects...)).Should(Succeed())
+			Eventually(func(g Gomega) bool {
+				var pipeline telemetryv1alpha1.TracePipeline
+				key := types.NamespacedName{Name: "pipeline-4"}
+				g.Expect(k8sClient.Get(ctx, key, &pipeline)).To(Succeed())
+				// pipeline-4 should become running after pipeline-1 is deleted
+				return pipeline.Status.HasCondition(telemetryv1alpha1.TracePipelineRunning)
+			}, timeout, interval).Should(BeTrue())
+		})
+	})
+
+	Context("When a broken tracepipeline exists", Ordered, func() {
+		BeforeAll(func() {
+			k8sObjects := makeTracingTestK8sObjects(portRegistry, "trace-mocks-broken-pipeline")
+			secondPipeline := makeBrokenTracePipeline("pipeline-2")
+
+			DeferCleanup(func() {
+				Expect(kitk8s.DeleteObjects(ctx, k8sClient, k8sObjects...)).Should(Succeed())
+				Expect(kitk8s.DeleteObjects(ctx, k8sClient, secondPipeline...)).Should(Succeed())
+			})
+			Expect(kitk8s.CreateObjects(ctx, k8sClient, k8sObjects...)).Should(Succeed())
+			Expect(kitk8s.CreateObjects(ctx, k8sClient, secondPipeline...)).Should(Succeed())
+		})
+
+		It("Should have running pipelines", func() {
+			Eventually(func(g Gomega) bool {
+				var pipeline telemetryv1alpha1.TracePipeline
+				key := types.NamespacedName{Name: "test"}
+				g.Expect(k8sClient.Get(ctx, key, &pipeline)).To(Succeed())
+				return pipeline.Status.HasCondition(telemetryv1alpha1.TracePipelineRunning)
+			}, timeout, interval).Should(BeTrue())
```

Comment:
Let's maybe extract this piece into the testkit since we repeat it many times here.

Reply:
I refactored the code a bit and removed the redundancy. Since this code also interacts with the K8s client, I would rather not move that into another package. Please take a look again.

Reply:
I moved the checks for running pipelines to a separate function, but kept it in this file to do not put the K8s client to testkit.